Replace legacy client in Docker Registry with external clientset

mfojtik opened the pull request:
This use the recent changes to codegen to provide missing clients for some of the image.openshift.io group resouces and then replace clients in registry with external clientsets.

In order to pickup external clients I had to rework the registry code in some places to deal with external types. I failed to do that in some places and fall-backed to conversions, which I hate but it will require more plumbing in the origin code to get this right.

Also I failed to use the external clientset for ImageSecrets as it is implemented as not-standard client returning `[]Secret` instead of `ImageSecretList` (so we can't generate client for that). 

Part of fixing: https://github.com/openshift/origin/issues/15272

mfojtik commented:
@smarterclayton surprise :-) PTAL pls.

mfojtik commented:
FYI, this is first compilable version, I still need to plumb tests. I haven't tested this live yet.

The clients will be merged before this: https://github.com/openshift/origin/pull/15625

mfojtik commented on pkg/dockerregistry/server/manifestservice.go:
```diff
@@ -152,15 +153,26 @@ func (m *manifestService) Put(ctx context.Context, manifest distribution.Manifes
+	// openshift helpers here, we need to fix this when we switch those to
+	// external versions.
+	image := imageapi.Image{}
+	if err := imageapiv1.Convert_v1_Image_To_image_Image(&ism.Image, &image, nil); err != nil {
```
@smarterclayton this is controversial... the FillImageMetadata is pretty heavy helper.

smarterclayton replied:
Why is having an external version bad?  They aren't changing that much.  A copy isn't that bad if tests validate nothing differs between two impls.  Image API helpers need to be moved out of that package anyway, and split into discrete groups.

mfojtik replied:
@smarterclayton i'm fine moving this into v1/helpers.go and start copying helpers we need externally there (including unit tests).

mfojtik commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -465,24 +460,37 @@ func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image,
+			context.GetLogger(r.ctx).Errorf("failed to convert image %s: %v", image.Name, err)
+			return
+		}
+		if err := imageapi.ImageWithMetadata(&newImage); err != nil {
```
should be util.ImageWithMetadata, will fix

legionus replied:
Can you use `util.ImageWithMetadata` here ?

mfojtik replied:
yeah, this is really just draft trying to gather early feedback

dmage commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -465,24 +460,29 @@ func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image,
 	}
 
 	if len(image.DockerImageLayers) > 0 {
+		newImage, err := util.ImageWithMetadata(image)
```
WIP?

mfojtik replied:
this will move to image/v1/helpers.go

smarterclayton commented on pkg/dockerregistry/server/util.go:
```diff
@@ -207,7 +209,82 @@ func (g *cachedImageStreamGetter) get() (*imageapi.ImageStream, error) {
 }
+
+// resolveImageID returns latest TagEvent for specified imageID and an error if
+// there's more than one image matching the ID or when one does not exist.
```
This is bad, and I don't think we should do it.  We want pkg/image/apis/image/helpers/v1 or something.

mfojtik replied:
@smarterclayton good idea

smarterclayton commented on pkg/image/generated/listers/image/internalversion/imagestreammapping.go:
```diff
@@ -0,0 +1,78 @@
+// This file was automatically generated by lister-gen
```
Split all these out into their own pull ahead of this Pr.

mfojtik replied:
already done, that commit will go away once this merge https://github.com/openshift/origin/pull/15625

stevekuznetsov commented:
/test extended_conformance_install_update

stevekuznetsov commented:
/test extended_conformance_install_update

stevekuznetsov commented:
/test extended_conformance_install_update

mfojtik commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (c *apiClient) ImageStreamSecrets(namespace string) ImageStreamSecretInterface {
+	return c.oc.ImageStreamSecrets(namespace)
```
@dmage `oc` is openshift client, `kube` is kubernetes client

dmage replied:
Don't we want to replace openshift client with clientset?

mfojtik replied:
@dmage we don't have client for ImageSecrets yet, so i need to keep legacy for that resource (for now). We need to fix the API to return correct resource instead of `[]Secret`.

mfojtik commented:
@smarterclayton @dmage @legionus now without any conversions \o/ i moved the helpers to v1/image/helpers.go (unit tests TBD).

miminar commented:
/test extended_image_registry

miminar commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,152 @@
+	userclientv1 "github.com/openshift/origin/pkg/user/generated/clientset/typed/user/v1"
+)
+
+// RegistryClient provides OpenShift and Kubernetes clients to Docker Registry.
```
Avoid product names.

miminar commented on pkg/dockerregistry/server/remoteblobgetter.go:
```diff
@@ -324,8 +326,11 @@ func identifyCandidateRepositories(
 			insecure := insecureByDefault
-			if tagRef, ok := is.Spec.Tags[tag]; ok {
-				insecure = insecureByDefault || tagRef.ImportPolicy.Insecure
+			for _, t := range is.Spec.Tags {
```
Why replacing key lookup (theoretically O(1)) with iteration O(n)? Did something happen to the keys?

mfojtik replied:
external vs. internal... fixing external to map will break the backward compatibility.

mfojtik commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,152 @@
+	config *clientcmd.Config
+}
+
+func NewRegistryClient(config *clientcmd.Config) RegistryClient {
```
I think I will stick with NewRegistryClient for now :)

mfojtik commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func (c *apiClient) LimitRanges(namespace string) LimitRangeInterface {
+	return c.kube.LimitRanges(namespace)
```
@deads2k or @smarterclayton do we have external clientset generated for this? this and the image secrets are the last remaining resources using 'legacy' (iow. internal) client.

mfojtik commented:
@miminar @smarterclayton @dmage the unit test should pass now, guess this is ready for review

miminar commented on pkg/dockerregistry/server/tagservice_test.go:
```diff
@@ -190,6 +191,16 @@ func TestTagCreation(t *testing.T) {
 			continue
 		}
 
+		r = newTestRepository(t, namespace, repo, testRepositoryOptions{
```
There's no need for re-instantiation.

mfojtik replied:
@dmage suggested this and it fixed the test case :-) still not sure why it was needed.

miminar commented on pkg/dockerregistry/testutil/fakeopenshift.go:
```diff
@@ -151,15 +154,35 @@ func (fos *FakeOpenShift) CreateImageStreamMapping(namespace string, ism *imagea
+				Image:                ism.Image.Name,
+			}},
+		})
+	}
```
:sob:

Looks like enterprise now.

dmage replied:
```
var tagEventList *imageapiv1.NamedTagEventList
for i, t := range is.Status.Tags {
	if t.Tag == ism.Tag {
		tagEventList = &is.Status.Tags[i]
		break
	}
}
if tagEventList == nil {
	is.Status.Tags = append(is.Status.Tags, imageapiv1.NamedTagEventList{
		Tag: ism.Tag,
	})
	tagEventList = &is.Status.Tags[len(is.Status.Tags)-1]
}

tagEventList.Items = append(tagEventList.Items, imageapiv1.TagEvent{
	DockerImageReference: ism.Image.DockerImageReference,
	Image:                ism.Image.Name,
})
```

mfojtik replied:
thx, this was quick and dirty hack :)

miminar commented on pkg/dockerregistry/testutil/fakeopenshift.go:
```diff
@@ -190,31 +213,48 @@ func (fos *FakeOpenShift) CreateImageStreamTag(namespace string, istag *imageapi
+	} else {
+		is.Status.Tags = append(is.Status.Tags, imageapiv1.NamedTagEventList{
+			Tag: imageTag,
+			Items: []imageapiv1.TagEvent{{
```
The tag event could be instantiated before the `if` to make it *dry*.

mfojtik replied:
reworked

miminar commented on pkg/image/apis/image/helper.go:
```diff
@@ -874,7 +874,8 @@ func UpdateTrackingTags(stream *ImageStream, updatedTag string, updatedImage Tag
 }
 
-func digestOrImageMatch(image, imageID string) bool {
+// DigestOrImageMatch matches the digest or image name.
```
… or their prefix

miminar commented on pkg/image/apis/image/v1/helpers.go:
```diff
@@ -0,0 +1,229 @@
+package v1
```
Could you vigorously prompt any potential editor to make the same changes to `pkg/image/apis/image/helper.go`?

mfojtik replied:
what changes? as we move to versioned objects also in origin, the internal helpers will be deprecated :)

miminar commented on pkg/dockerregistry/server/pullthroughblobstore_test.go:
```diff
@@ -624,6 +626,7 @@ func TestPullthroughServeBlobInsecure(t *testing.T) {
 
 		_, err = ptbs.Stat(ctx, dgst)
 		if err != tc.expectedStatError {
+			t.Logf("blob: %s", tc.blobDigest)
```
debugging leftover?

dmage commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -166,11 +167,11 @@ type repository struct {
 	repo distribution.Repository,
 	options map[string]interface{},
 ) (distribution.Repository, error) {
-	// TODO: Deprecate this environment variable.
```
TODO is no more valid?

dmage commented on pkg/dockerregistry/testutil/fakeopenshift.go:
```diff
@@ -151,15 +154,35 @@ func (fos *FakeOpenShift) CreateImageStreamMapping(namespace string, ism *imagea
 
 	if is.Status.Tags == nil {
-		is.Status.Tags = make(map[string]imageapi.TagEventList)
+		is.Status.Tags = []imageapiv1.NamedTagEventList{}
```
You can remove this `if` completely.

dmage commented on pkg/dockerregistry/testutil/fakeopenshift.go:
```diff
@@ -190,31 +213,48 @@ func (fos *FakeOpenShift) CreateImageStreamTag(namespace string, istag *imageapi
 
 	if is.Spec.Tags == nil {
-		is.Spec.Tags = make(map[string]imageapi.TagReference)
+		is.Spec.Tags = []imageapiv1.TagReference{}
```
Can be removed.

mfojtik commented on pkg/dockerregistry/server/auth_test.go:
```diff
@@ -416,13 +440,13 @@ func TestAccessController(t *testing.T) {
 		}
 		if !reflect.DeepEqual(actions, &expectedActions) {
-			t.Errorf("%s: expected\n\t%#v\ngot\n\t%#v", k, &expectedActions, actions)
+			t.Errorf("\n%s:\n expected:\n\t%#v\ngot:\n\t%#v", k, &expectedActions, actions)
```
@miminar but it will increase the number of packages we import from kube into registry :-) not sure if it is worth.

mfojtik commented on pkg/dockerregistry/server/prune/prune.go:
```diff
@@ -64,9 +65,15 @@ func Prune(ctx context.Context, storageDriver driver.StorageDriver, registry dis
 		// Keep the manifest.
 		inuse[image.Name] = image.DockerImageReference
 
+		if err := imageapiv1.ImageWithMetadata(&image); err != nil {
```
@miminar yes

mfojtik commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -135,11 +136,11 @@ type repository struct {
 	registryAddr     string
 	namespace        string
 	name             string
+	config           *configuration.Configuration
```
this is @legionus code for metrics... maybe we should just pass a boolean indicating if metrics are enabled.

mfojtik commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -396,7 +391,7 @@ func (r *repository) createImageStream(ctx context.Context) (*imageapi.ImageStre
 
 // getImage retrieves the Image with digest `dgst`. No authorization check is done.
-func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
+func (r *repository) getImage(dgst digest.Digest) (*imageapiv1.Image, error) {
```
sure, done :)

mfojtik commented:
@deads2k or @enj 

```
I0815 15:19:26.458630    1568 wrap.go:42] POST /apis/authorization.k8s.io/v1/namespaces/integration/localsubjectaccessreviews: (1.719261ms) 422 [[integration.test/v1.7.0+695f48a16f (linux/amd64) kubernetes/d2e5420] 127.0.0.1:59492]
time="2017-08-15T15:19:26.458897159Z" level=error msg="OpenShift client error: .authorization.k8s.io \"\" is invalid: spec.user: Invalid value: \"\": at least one of user or group must be specified" go.version=go1.8.3 http.request.host="127.0.0.1:5000"
```

guess the user field is mandatory for localSAR ?

mfojtik commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -342,7 +342,7 @@ func GetOpenshiftBootstrapClusterRoles() []authorizationapi.ClusterRole {
 				authorizationapi.NewRule("create").Groups(authzGroup, legacyAuthzGroup).Resources("localresourceaccessreviews", "localsubjectaccessreviews", "subjectrulesreviews").RuleOrDie(),
 				authorizationapi.NewRule("create").Groups(securityGroup, legacySecurityGroup).Resources("podsecuritypolicysubjectreviews", "podsecuritypolicyselfsubjectreviews", "podsecuritypolicyreviews").RuleOrDie(),
-				authorizationapi.NewRule("create").Groups(kAuthzGroup).Resources("localsubjectaccessreviews").RuleOrDie(),
+				authorizationapi.NewRule("create").Groups(kAuthzGroup).Resources("localsubjectaccessreviews", "selfsubjectaccessreviews").RuleOrDie(),
```
@deads2k check :)

mfojtik commented on pkg/dockerregistry/server/auth.go:
```diff
@@ -489,17 +450,19 @@ func verifyOpenShiftUser(ctx context.Context, client client.UsersInterface) (str
+	sar := authorizationapi.SelfSubjectAccessReview{
+		Spec: authorizationapi.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Namespace: namespace,
```
@deads2k check :-) seems to fix the integration test

deads2k replied:
Yes, this matches expectations.

mfojtik commented:
/test extended_conformance_gce

flake:  https://github.com/openshift/origin/issues/15792

mfojtik commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -891,7 +891,6 @@ func GetOpenshiftBootstrapClusterRoles() []authorizationapi.ClusterRole {
 				authorizationapi.NewRule("get", "update").Groups(imageGroup, legacyImageGroup).Resources("imagestreams/layers").RuleOrDie(),
 				authorizationapi.NewRule(readWrite...).Groups(authzGroup, legacyAuthzGroup).Resources("rolebindings", "roles").RuleOrDie(),
 				authorizationapi.NewRule("create").Groups(authzGroup, legacyAuthzGroup).Resources("localresourceaccessreviews", "localsubjectaccessreviews", "subjectrulesreviews").RuleOrDie(),
-				authorizationapi.NewRule("create").Groups(kAuthzGroup).Resources("localsubjectaccessreviews").RuleOrDie(),
```
@deads2k why was this here? registry used legacy client before, not hitting the k8s authz... guess we can remove this without concerns about backward compatibility?

mfojtik commented:
/retest

mfojtik commented:
@miminar @dmage i might need some help figuring out what is causing the last 2 test failures... the manifest seems to be still unknown even though i fixed the fill image with metadata :/

mfojtik commented:
@miminar @dmage the registry logs seems to return:

```
time="2017-08-16T20:05:16.311476196Z" level=error msg="image does not have metadata sha256:8573b4a813d7b90ef3876c6bec33db1272c02f0f90c406b25a5f9729169548ac" go.version=go1.8.3 instance.id=d5a2f5f0-9ff2-4592-866c-c64d72dbf187 openshift.logger=registry 
```

mfojtik commented:
test/cmd flake is: https://github.com/openshift/origin/issues/15900

legionus commented on pkg/image/apis/image/v1/helpers.go:
```diff
@@ -0,0 +1,228 @@
+			image.DockerImageLayers[i].MediaType = layer.MediaType
+		}
+		// reverse order of the layers for v1 (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
```
I guess this is wrong. [The layer list is ordered starting from the base image (opposite order of schema1)](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#image-manifest-field-descriptions)

legionus commented on pkg/image/apis/image/v1/helpers.go:
```diff
@@ -113,10 +113,12 @@ func ImageWithMetadata(image *Image) error {
 	case 1:
 		image.DockerImageManifestMediaType = schema1.MediaTypeManifest
 
-		if len(manifest.History) == 0 {
```
@mfojtik why ? This code should not change.

mfojtik replied:
@legionus was just an attempt :-)

mfojtik replied:
@legionus i'm still trying to figure out the problem with invalid manifest... unfortunately i haven't fixed it yet :/

mfojtik replied:
@legionus here: https://github.com/openshift/origin/pull/15624/files#diff-1a18cb8c0d7684cefcef809b9d01aac3R411

we getting this:

```
 image layers: 1, manifest: 0. media type: application/vnd.docker.distribution.manifest.v2+json <nil>" go.version=go1.8.3 instance.id=2226bb84-d159-43e3-9fd2-d3c43ea0a1a7 openshift.logger=registry
```

It seems like the manifest is empty/missing, but the image has 1 layer and the media type... no idea why.

mfojtik replied:
@legionus the manifest check at the top of ImageWithMetadata will return `nil` in that case so we don't have the metadata... I wonder why the manifest is empty in this case.

mfojtik commented:
Thanks to @legionus we found out why the ImageWithMetadata was not filling in the ImageMetaData field.

Since the external representation of an Image object does not have manifest, the ImageWithMetadata was basically useless. Fortunately, when we serialize the internal version we fill in the DockerImageMetadata.Raw with JSON representation of DockerImage AND we also populate the layers. So instead of copying what the internal version of ImageWithMetadata does we just need to decode that JSON into DockerImageMetadata.Object and then assert it properly in the registry code.

The end-to-end should pass now, so I guess this is ready for "final" review. @smarterclayton @miminar @legionus @dmage

mfojtik commented:
@legionus this is now failing even worse than before ;-)))

mfojtik commented:
@legionus it seems like the fix broke how we store the image:

```
[mfojtik@dev ~]$ oc get isimage -o yaml busybox@sha256:2605a2c4875ce5eb27a9f7403263190cd1af31e48a2044d400320548356251c4
apiVersion: v1
image:
  dockerImageLayers: null
  dockerImageManifestMediaType: application/vnd.docker.distribution.manifest.v2+json
  dockerImageMetadata:
    ContainerConfig: {}
    Created: null
    Id: ""
    apiVersion: "1.0"
    kind: DockerImage
  dockerImageMetadataVersion: "1.0"
  dockerImageReference: 172.30.176.20:5000/cache/busybox@sha256:2605a2c4875ce5eb27a9f7403263190cd1af31e48a2044d400320548356251c4
```

This can be reproduced easely by docker pushing the image into registry. Any image created looks like this...

mfojtik commented:
@legionus OMG almost green :-)

legionus commented:
> can you please check the last e2e failure and dig into why we return v1 manifest instead of v2 manifest?

@mfojtik ok. I will.

mfojtik commented:
flake: https://github.com/openshift/origin/issues/15912

/test extended_conformance_install_update

mfojtik commented:
@legionus I think I fixed the last problem, the DockerConfig has to be populated from the manifest blob in FillMetadata for schema2 in order to set the proper media type. I moved `imageMetadataFromManifest` inside `FillImageMetadata` after we get DockerConfig and that fixed the problem for me locally.

The test should now be green.

mfojtik commented on pkg/dockerregistry/server/util.go:
```diff
@@ -207,7 +214,176 @@ func (g *cachedImageStreamGetter) get() (*imageapi.ImageStream, error) {
+// imageMetadataFromManifest is used only when creating the image stream mapping
+// in registry. In that case the image stream mapping contains image with the
+// manifest and we have to pupulate the the docker image metadata field.
+func imageMetadataFromManifest(image *imageapiv1.Image) error {
```
@smarterclayton we moved this to registry as this function is only used by registry (when you create ISM and need to place the metadata in RawExtension as registry use external client).

deads2k replied:
> @smarterclayton we moved this to registry as this function is only used by registry (when you create ISM and need to place the metadata in RawExtension as registry use external client).

Moved?  Is this something that our API server did before?  Does the API server still do it if it is missing?  Recall that during an upgrade, we run skewed apiserver to registry.

smarterclayton replied:
Moved or copied?  I don't see a corresponding delete.

mfojtik replied:
@smarterclayton moved in terms of this PR. This helper lived in v1/helper.go before (in this PR). We realized that the only user of this function is registry itself when we need to convert the Image in ISM to JSON (so when we create the ISM.

@deads2k we did this before, but we had a conversions that convert the external/internal image and fill-in all the fields we required... I don't like this pattern TBH (that external vs. internal use JSON vs. regular fields). I believe this was to support multiple versions of manifest/docker config. @miminar ?

mfojtik commented:
The failure seems to be pulp?

```
I0825 10:07:40.582038   29867 client.go:603] unable to find tagged image at "https://registry.access.redhat.com/v2/library/rhel/manifests/latest", 404 map[Content-Type:[application/json] Date:[Fri, 25 Aug 2017 10:07:40 GMT] Server:[Apache] Content-Length:[53]]: {"errors": [{"code": "404", "message": "Not Found"}]}
```

mfojtik commented:
/test integration

mfojtik commented:
/test integration

mfojtik commented:
/test extended_conformance_gce

deads2k commented on pkg/dockerregistry/server/auth_test.go:
```diff
@@ -19,16 +19,26 @@ import (
 	"github.com/docker/distribution/context"
 	"github.com/openshift/origin/pkg/api/latest"
-	authorizationapi "github.com/openshift/origin/pkg/authorization/apis/authorization"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
```
Eventually this dependency will need to snipped. Not a blocker.

deads2k commented on pkg/dockerregistry/server/auth_test.go:
```diff
@@ -19,16 +19,26 @@ import (
+func sarResponse(ns string, allowed bool, reason string) *authorizationapi.SelfSubjectAccessReview {
+	resp := &authorizationapi.SelfSubjectAccessReview{}
+	if len(ns) > 0 {
+		resp.Namespace = ns
```
No need to be conditional.

deads2k commented on pkg/dockerregistry/server/auth_test.go:
```diff
@@ -416,13 +440,13 @@ func TestAccessController(t *testing.T) {
 		}
 		if !reflect.DeepEqual(actions, &expectedActions) {
-			t.Errorf("%s: expected\n\t%#v\ngot\n\t%#v", k, &expectedActions, actions)
+			t.Errorf("\n%s:\n expected:\n\t%#v\ngot:\n\t%#v", k, &expectedActions, actions)
```
> @miminar but it will increase the number of packages we import from kube into registry :-) not sure if it is worth.

Every project is allowed to depend on `k8s.io/apimachinery`.

deads2k commented on pkg/dockerregistry/server/blobdescriptorservice.go:
```diff
@@ -208,17 +210,10 @@ func imageHasBlob(
-			// has already been done. There are none.
-			return false
-		}
-		err = imageapi.ImageWithMetadata(image)
```
Dropping this `ImageWithMetadata` isn't obvious to me.

@miminar ptal.

mfojtik replied:
@deads2k this is more complicated... external Image have everything encoded in JSON inside DockerImageMetadata so the filling happens differently there...

deads2k commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,162 @@
+
+// Interface contains client methods that registry use to communicate with
+// Origin or Kubernetes API.
+type Interface interface {
```
Interesting.  I may not have done it, but we'll see how it plays out.

mfojtik replied:
@deads2k we needed a wrapper if we want to potentially add metrics for the API calls, I think that was originally @legionus idea

deads2k commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,162 @@
+
+	return newAPIClient(
+		oc,
+		kc.Core(),
```
why aren't these constructed like the others?

mfojtik replied:
agree, we don't have force us to use clientmd

deads2k commented on pkg/dockerregistry/server/client/client.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// safeClientConfig returns a client config without authentication info.
+func (с *registryClient) safeClientConfig() restclient.Config {
```
eliminate this function and call the `AnonymousClientConfig` directly.  It's more obvious for people familiar with kube patterns.

deads2k commented on pkg/dockerregistry/server/client/interfaces.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+type ImageSignatureInterface interface {
+	Create(signature *imageapiv1.ImageSignature) (*imageapiv1.ImageSignature, error)
```
Need type assertions to make sure that the generated clients conform to these interfaces.

mfojtik replied:
+1 good catch

deads2k commented on pkg/dockerregistry/server/client/test.go:
```diff
@@ -0,0 +1,31 @@
+package client
```
followup issue: please move this to a `fake` package.

deads2k commented on pkg/dockerregistry/server/manifestschema1handler.go:
```diff
@@ -51,7 +52,7 @@ type manifestSchema1Handler struct {
 var _ ManifestHandler = &manifestSchema1Handler{}
 
-func (h *manifestSchema1Handler) FillImageMetadata(ctx context.Context, image *imageapi.Image) error {
+func (h *manifestSchema1Handler) FillImageMetadata(ctx context.Context, image *imageapiv1.Image) error {
```
pre-existing, but mutation of args always rubs me the wrong way.

mfojtik replied:
will capture this as follow up for @legionus

deads2k commented on pkg/dockerregistry/server/repositorymiddleware.go:
```diff
@@ -135,11 +135,12 @@ type repository struct {
 	registryAddr     string
 	namespace        string
 	name             string
+	//config           *configuration.Configuration
```
cruft

deads2k commented:
one question about backwards compatibility.  Otherwise, if it looks good to the registry guys, it looks good to me.

/approve

mfojtik commented:
@deads2k addressed comments, I also refactored the client a little to be more DRY. Do you know if there is an alternative for `pkg/cmd/util/clientcmd` ? (guess we need to read env vars/etc.)

mfojtik commented:
/test extended_image_registry

legionus commented on pkg/dockerregistry/server/util.go:
```diff
@@ -207,7 +214,176 @@ func (g *cachedImageStreamGetter) get() (*imageapi.ImageStream, error) {
+			image.DockerImageLayers[i].MediaType = layer.MediaType
+		}
+		// reverse order of the layers for v1 (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
```
@mfojtik I guess this is wrong. [The layer list is ordered starting from the base image (opposite order of schema1)](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#image-manifest-field-descriptions)

mfojtik replied:
@legionus i just copied this over from the original helper (I guess it is broken there as well). Mind fixing this as a follow up issue? (don't think it is a blocker for this PR).

legionus replied:
@mfojtik Sure. Your PR is so huge already.

legionus commented:
/lgtm
/approve

mfojtik commented on pkg/dockerregistry/server/blobdescriptorservice.go:
```diff
@@ -15,6 +15,7 @@ import (
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	imageapiv1 "github.com/openshift/origin/pkg/image/apis/image/v1"
```
agree, we should probably copy the docker types into external version, but I hate to do that just to get rid of this import (we will have two relatively large structs to keep eye on).

legionus replied:
@mfojtik It's not about imports, but about the self-sufficiency of API. In this PR to read the metadata received from the external API you have to use internal API. This means that the external API is not complete.

mfojtik commented:
/test extended_conformance_install_update

mfojtik commented:
/test extended_networking_minimal

mfojtik commented:
flake: https://github.com/openshift/origin/issues/16013

/test extended_conformance_gce

mfojtik commented:
@legionus @deads2k \o/

legionus commented:
@mfojtik congrats! ))
